vhost-device-console: Add initial implementation

TimosAmpel opened the pull request:
### Summary of the PR

This PR introduces an initial implementation of vhost-device-console device.
The development of this device was performed by Virtual Open Systems in the context
of the [Automotive Grade Linux](https://www.automotivelinux.org/) SDV-EG (Software Defined Vehicle expert group)
AGL Native VIRTIO project. Part of this development has already been upstreamed
(see [AGL Gerrit console patch](https://gerrit.automotivelinux.org/gerrit/c/AGL/meta-agl-devel/+/29545)).

The new device creates a console device which communicates with vhost-user-console
frontend driver and eventually with virtio-console. Since there was no upstream
implementation of any vhost-user-console frontend in QEMU, the current version has
only been tested with virtio-loopback design.

Virtio-loopback architecture aims to create a new HAL (hardware abstraction layer),
designed for non-Hypervisor environments based on virtio. The main objective
is to enable virtio drivers to communicate with vhost-user devices without emulation
or virtualization being involved. This technology has also been upstreamed
into [AGL gerrit repositories](https://gerrit.automotivelinux.org/gerrit/c/AGL/meta-agl-devel/+/28257).

More information about virtio-loopback can be found in the links below:
- [Virtio-loopback design document](https://git.virtualopensystems.com/virtio-loopback/docs/-/blob/epsilon-release/design_docs/EG-VIRT_VOSYS_virtio_loopback_design_v1.4_2023_04_03.pdf)
- [How to test the technology](https://git.virtualopensystems.com/virtio-loopback/docs/-/blob/epsilon-release/README.md)

Since we believe that this device might be interesting to the community,
this PR targets to share this work, ask for comments and potentially
merge the implementation under vhost-device repository.

### Requirements

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

epilys commented on staging/vhost-device-console/src/backend.rs:
```diff
@@ -0,0 +1,240 @@
+struct ConsoleArgs {
+    /// Location of vhost-user Unix domain socket. This is suffixed by 0,1,2..socket_count-1.
+    #[clap(short = 's', long)]
+    socket_path: String,
+
+    /// Location of Unix domain socket which all virtio-console traffic will be routed
+    #[clap(short = 'p', long, default_value = "path_to_console_socket")]
+    console_path: String,
+
+    /// Number of guests (sockets) to connect to.
+    #[clap(short = 'c', long, default_value_t = 1)]
+    socket_count: u32,
+}
+
+#[derive(PartialEq, Debug)]
+struct ConsoleConfiguration {
+    socket_path: String,
+    socket_count: u32,
+    console_path: String,
+}
```
It's better to use `PathBuf` and `&Path` than `String`. You'll have access to all the path APIs and you will have path value validation "for free" by Clap's parser; no need to do white space trimming or other sanitizing operations to console arguments.

epilys commented on staging/vhost-device-console/src/backend.rs:
```diff
@@ -0,0 +1,240 @@
+}
+
+fn start_backend(args: ConsoleArgs) -> Result<()> {
+
```
```suggestion
```

epilys commented on staging/vhost-device-console/src/backend.rs:
```diff
@@ -0,0 +1,240 @@
+    let config = ConsoleConfiguration::try_from(args).unwrap();
+    let mut handles = Vec::new();
+
+    for _ in 0..config.socket_count {
+        let socket = config.socket_path.to_owned();
+        let console_path = config.console_path.to_owned();
```
We refactored some other devices from `String` to `PathBuf` for path arguments recently, feel free to copy the code handling socket paths and suffixing the integer count.

epilys commented on staging/vhost-device-console/src/backend.rs:
```diff
@@ -0,0 +1,240 @@
+                .set_vring_worker(&vring_workers[0]);
+            let read_handle = VhostUserConsoleBackend::start_console_thread(&vu_console_backend);
+
+            let listener = Listener::new(socket.clone(), true).unwrap();
```
I think you meant to use `SOCKET_PATH{0,1,..., socket_count}` here instead of the generic `socket` path, right?

epilys commented on staging/vhost-device-console/src/backend.rs:
```diff
@@ -0,0 +1,240 @@
+        handles.push(handle);
+    }
+
+    for handle in handles {
+        handle.join().map_err(|_| Error::FailedJoiningThreads)??;
+    }
```
We discovered this leads to problems when a thread dies but the code here is blocked on a thread handle on a smaller index in `handles`.

See a better approach here using channels to signal thread shutdown: https://github.com/rust-vmm/vhost-device/blame/5a2bd27037ef1efdcfdffa733e8c34c25204d6e5/vhost-device-input/src/main.rs#L125-L161

epilys commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,754 @@
+                trace!("VIRTIO_CONSOLE_DEVICE_READY");
+                self.ready = true;
+                ctrl_msg_reply.event = VIRTIO_CONSOLE_PORT_ADD.into();
+                self.rx_fifo.add(ctrl_msg_reply).unwrap();
```
If `self.rx_fifo.add` failing is fatal enough to panic with unwrap instead of returning an error (for example with `ok_or_else`) maybe we can convert the unwrap to `expect` and write the panic reason.

epilys commented on staging/Cargo.toml:
```diff
@@ -2,4 +2,5 @@
 resolver = "2"
 members = [
   "vhost-device-video",
+  "vhost-device-console",
```
```suggestion
  "vhost-device-console",
  "vhost-device-video",
```

TimosAmpel commented:
> Hi @TimosAmpel , thanks for this PR!
> 
> I'll review next week, but in the meantime I'd like to understand better `virtio-loopback`. Is it similar to VDUSE? Is there any RFC or patch posted upstream in the Linux kernel?
> 
> That said, QEMU 8.2 should support a generic vhost-user device, so maybe you give it a try to test this device using something like this: (@stsquad and @epilys knows better than me)
> 
> ```
> -device vhost-user-device-pci,virtio-id=41,num_vqs=2,config_size=8
> ```

Hi @stefano-garzarella, thanks for you response.

Virtio-loopback is indeed similar to VDUSE and the main differences between
them are mainly: 
1) the data sharing mechanism: virtio-loopback implements a zero-copy memory
   mapping mechanism, the data are directly accessible by the user-space.
   VDUSE is based on a bouncing buffer mechanism which does not implement
   the zero-copy principle.
2) virtio-loopback is not implemented on top of vDPA, aiming for a simpler
   solution.

In addition, virtio-loopback supports the following rust-vmm devices:
vhost-device-gpio, input, rng, sound.  

An RFC presenting the architecture (no patches included) had been submitted to
the upstream virtio-dev mailing list, please check the following [link](https://lists.oasis-open.org/archives/virtio-dev/202304/msg00324.html).
The upstream implementation of the driver and adapter (user-space component)
can be found in the AGL repositories [link1](https://gerrit.automotivelinux.org/gerrit/gitweb?p=src/virtio/virtio-loopback-driver.git;a=summary) and [link2](https://gerrit.automotivelinux.org/gerrit/gitweb?p=src%2Fvirtio%2Fvirtio-loopback-adapter.git;a=summary) correspondingly.

Thank you a lot for the "vhost-user-device-pci" suggestion, I plan to test both
vhost-device-console and vhost-device-can devices with it and apply any changes
if needed.

TimosAmpel commented:
Hello @epilys and @stefano-garzarella,

I have pushed a new simplified version of the device
with all your comments integrated.

Vhost-device-console is being tested with QEMU's "vhost-user-device" as suggested:
```
qemu-system-x86_64  \
        -m 4096 \
        -numa node,memdev=mem \
        -object memory-backend-file,id=mem,size=4G,mem-path=/dev/shm,share=on \
        -chardev socket,id=char1,path=/tmp/console.sock \
        -device vhost-user-device-pci,chardev=char1,virtio-id=3,num_vqs=4,config_size=12 \
```

Also, I have created a new QEMU `vhost-user-console` front-end based on 
`vhost-user-device` which can be found in this [repository](ihttps://github.com/virtualopensystems/vhost-device/tree/vhost-device-console-rfc).

More about, known limination, test examples and future work can be found in
the README file.

epilys commented on staging/vhost-device-console/src/console.rs:
```diff
@@ -0,0 +1,63 @@
+use log::trace;
+use thiserror::Error as ThisError;
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Copy, Clone, Debug, PartialEq, ThisError)]
+pub(crate) enum Error {}
+
+#[derive(Debug)]
+pub(crate) struct ConsoleController {
+    config: VirtioConsoleConfig,
+    pub exit: bool,
+}
+
+impl ConsoleController {
+    // Creates a new controller corresponding to `device`.
+    pub(crate) fn new() -> Result<ConsoleController> {
+        Ok(ConsoleController {
+            config: VirtioConsoleConfig {
+                cols: 20.into(),
+                rows: 20.into(),
+                max_nr_ports: 1.into(),
+                emerg_wr: 64.into(),
+            },
+            exit: false,
+        })
+    }
```
Is returning a result a leftover from previous iterations? Since `new` cannot fail it's redundant.

TimosAmpel replied:
Thanks, I will fix that on the next push!

epilys commented on staging/vhost-device-console/src/console.rs:
```diff
@@ -0,0 +1,63 @@
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_console_controller_creation() {
+        let controller = ConsoleController::new();
+        assert!(controller.is_ok());
+    }
+
```
Generally results should be `unwrap`ped or `expect`ed in tests, so that their error message shows up in the test results. Here though you already unwrap (the infallible as mentioned above) `ConsoleController::new()` in the `test_console_controller_config` so `test_console_controller_creation` is unnecessary.

TimosAmpel commented:
Hello @epilys,

I have pushed a newer version which covers the scenario of the device working with multiple QEMU's guest 
and two different modes (backends: network console, nested console).

As before, the readme file has been updated and registers any known limitation, new features, testing examples etc.

stefano-garzarella commented:
@TimosAmpel if this PR is not a Draft, I suggest cleaning a bit the commits and squash them. We usually prefer to merge a PR with commits that do not modify code added in the same PR.

TimosAmpel commented:
> @TimosAmpel if this PR is not a Draft, I suggest cleaning a bit the commits and squash them. We usually prefer to merge a PR with commits that do not modify code added in the same PR.

Thanks @stefano-garzarella! I have squashed them all in one new presenation commit, I hope that's more suitable.

stefano-garzarella commented:
@TimosAmpel can you rebase this PR, so next week I'll review it again ;-)

dorindabassey commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1436 @@
+const TX_QUEUE: u16 = 1;
+const CTRL_RX_QUEUE: u16 = 2;
+const CTRL_TX_QUEUE: u16 = 3;
+const BACKEND_RX_EFD: u16 = (NUM_QUEUES + 1) as u16;
+const BACKEND_CTRL_RX_EFD: u16 = (NUM_QUEUES + 2) as u16;
```
I think it will be nice to have some documentation comment for these queues given that it's not in the list of virtqueues in the spec, so we know what it's usage.

dorindabassey commented on staging/vhost-device-console/Cargo.toml:
```diff
@@ -0,0 +1,40 @@
+epoll = "4.3"
+libc = "0.2"
+log = "0.4"
+rand = "0.8.5"
+tempfile = "3.9"
```
```suggestion
```

dorindabassey replied:
these dependencies are unused

dorindabassey commented on staging/vhost-device-console/src/console.rs:
```diff
@@ -0,0 +1,48 @@
+use crate::vhu_console::VirtioConsoleConfig;
+use clap::ValueEnum;
+use log::trace;
+use thiserror::Error as ThisError;
+
+#[derive(Copy, Clone, Debug, PartialEq, ThisError)]
+pub(crate) enum Error {}
```
```suggestion
```

dorindabassey replied:
unused definition

dorindabassey commented on staging/vhost-device-console/src/backend.rs:
```diff
@@ -0,0 +1,363 @@
+            .unwrap()
+            .set_vring_worker(&vring_workers[0]);
+
+        // Start the corresponinding console thread
```
```suggestion
        // Start the corresponding console thread
```

dorindabassey commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1436 @@
+            return Err(Error::UnexpectedReadableDescriptor(0));
+        }
+
+        //let response = self.input_buffer.clone();
```
```suggestion
```

TimosAmpel commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1436 @@
+
+            if ctrl_msg.event.to_native() == VIRTIO_CONSOLE_PORT_NAME {
+                let string_bytes = "org.test.foo!".as_bytes();
+                buffer.extend_from_slice(string_bytes);
```
Hello @dorindabassey, thanks a lot for your comments! I think I should have addressed them all till this one.

dorindabassey commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1441 @@
+                trace!("VIRTIO_CONSOLE_DEVICE_READY");
+                self.ready = true;
+                ctrl_msg_reply.event = VIRTIO_CONSOLE_PORT_ADD.into();
+                self.rx_ctrl_fifo
+                    .add(ctrl_msg_reply)
+                    .expect("Control message could not be added into queue");
```
is it intended to panic here? can the device not recover from this error? 
if that's not the case, I will suggest to use the `map_err` method to properly propagate the error.

dorindabassey commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1441 @@
+                                            let input_buffer =
+                                                String::from_utf8_lossy(&buffer[..bytes_read])
+                                                    .to_string();
+                                            let _ = vhu_console
+                                                .write()
+                                                .unwrap()
+                                                .rx_data_fifo
+                                                .add(input_buffer);
```
```suggestion
                                            vhu_console
                                                .write()
                                                .unwrap()
                                                .rx_data_fifo
                                                .add(input_buffer).unwrap();
```

dorindabassey commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1441 @@
+                            }
+
+                            if fdset_clone.contains(term.as_raw_fd()) {
+                                let character =
+                                    match term.read_key().expect("Error reading character!") {
+                                        Key::Char(character) => Some(character),
+                                        Key::Enter => Some('\n'),
+                                        Key::Tab => Some('\t'),
+                                        Key::Backspace => Some('\u{8}'),
+                                        _ => None,
+                                    };
+
+                                if character.is_some() {
+                                    // Pass the data to vhu_console and trigger an EventFd
+                                    let input_buffer =
+                                        character.expect("Unexpected character!").to_string();
+                                    let _ =
+                                        vhu_console.write().unwrap().rx_data_fifo.add(input_buffer);
+                                    vhu_console.write().unwrap().rx_event.write(1).unwrap();
+                                }
```
```suggestion
                                if let Some(character) = match term.read_key().unwrap() {
                                    Key::Char(character) => Some(character),
                                    Key::Enter => Some('\n'),
                                    Key::Tab => Some('\t'),
                                    Key::Backspace => Some('\u{8}'),
                                    _ => None,
                                } {
                                    // Pass the data to vhu_console and trigger an EventFd
                                    let input_buffer = character.to_string();
                                    vhu_console
                                        .write()
                                        .unwrap()
                                        .rx_data_fifo
                                        .add(input_buffer)
                                        .unwrap();
                                    vhu_console.write().unwrap().rx_event.write(1).unwrap();
                                }
```
I'd prefer this

TimosAmpel replied:
@dorindabassey, thanks for the review comments. I think all of them have been addressed. Let me know if you have any additional comments to discuss

stefano-garzarella commented on staging/vhost-device-console/CHANGELOG.md:
```diff
@@ -0,0 +1,15 @@
+
+### Deprecated
+
+## [0.1.0]
```
We haven't done the release yet, so for now I wouldn't put anything in the Changelo, just the Unreleased template.

stefano-garzarella replied:
Let' keep v0.1.0 in the changelog, since we will do the release after merging this.
BTW can you fix the style following: https://github.com/rust-vmm/vhost-device/pull/695

stefano-garzarella commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,131 @@
+This program is tested with QEMU's `vhost-user-device-pci` device.
+Examples' section below.
+
+This program is a vhost-user backend that emulates a VirtIO Console device.
```
Duplicated line?

TimosAmpel replied:
Yes indeed, I have deleted it!

stefano-garzarella commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,131 @@
+terminal.
+
+By using "network" backend, vhost-device-console, as soon as the guests
+is booted, creates a local server in the given localhost port, and gives
```
Which kind of connection? tcp?

TimosAmpel replied:
Yes that's a TCP connection. The device has a TCPlisener when started with `network` option, and waits for a TCP connection.
I describe that later in the README file but I have also added a clarification where you pointed out.

stefano-garzarella reviewed (commented):
Sorry for the delay. Can you rebase this work?

TimosAmpel commented:
Hello @stefano-garzarella,

I have pushed a newer version in which I have integrated your latests suggestions and also the ones you provided on [PR 602 - CAN](https://github.com/rust-vmm/vhost-device/pull/602). Those include:
- Re-arranging the `main.rs` file to include Args related functions
- Creating a new file `virtio_console.rs` which hosts all defines and most of the structures used
- Modify the CHANGELOG, Cargo.toml, README files
- Rebase the code on latest main branch

Please feel free to provide any additional comment!

stefano-garzarella commented:
> I have pushed a newer version in which I have integrated your latests suggestions and also the ones you provided on 
[PR 602 - CAN](https://github.com/rust-vmm/vhost-device/pull/602). 

@TimosAmpel great! This is really much appreciated!

> Please feel free to provide any additional comment!

I'll do next week ;-)

stefano-garzarella commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+
+/// Feature bit numbers
+#[allow(dead_code)]
+pub const VIRTIO_CONSOLE_F_SIZE: u16 = 0;
+#[allow(dead_code)]
+pub const VIRTIO_CONSOLE_F_MULTIPORT: u16 = 1;
+pub const VIRTIO_CONSOLE_F_EMERG_WRITE: u16 = 2;
```
Should we move them to `virtio_console.rs`?

TimosAmpel replied:
Good idea! They are moved into `virtio_console.rs`.

stefano-garzarella reviewed (commented):
After few tries I had this:

```
thread 'vring_worker' panicked at vhost-device-console/src/vhu_console.rs:235:55:
called `Result::unwrap()` on an `Err` value: FromUtf8Error { bytes: [239], error: Utf8Error { valid_up_to: 0, error_len: None } }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at vhost-device-console/src/vhu_console.rs:560:60:
called `Result::unwrap()` on an `Err` value: PoisonError { .. }
```

Also I'm not sure if it's releated to the device or the driver, but with telnet or nc, it's not working very well:

```
[root@f38-vm-build ~]# echo hello
[root@f38-vm-build ~]# echo hello

[root@f38-vm-build ~]# echo hello


[root@f38-vm-build ~]# echo hello
echo hello
[root@f38-vm-build ~]# 
```

Is that expected? Or am I doing something wrong?

MatiasVara commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,129 @@
+
+This program is tested with QEMU's `vhost-user-device-pci` device.
+Examples' section below.
+
```
In case of CAN, for example, I think we can say that we are waiting for the spec and for the driver to be accepted upstream.

stefano-garzarella replied:
@TimosAmpel can you add a section in this README to explain why this device is in staging? I'd describe the missing steps, something like this (not sure if they apply in this case, it's just an example):
```markdown
## Staging Device
This device will be in `staging` until we complete the following steps:
- [ ] merge Linux's driver upstream
- [ ] merge virtio-spec changes upstream
- [ ] ...
```

MatiasVara commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+                print!("{}", my_string);
+                io::stdout().flush().unwrap();
+            } else {
+                self.output_buffer = my_string.clone();
```
I am probably missing something here but can you explain what this line does? Is it copying the content of each descriptor to `output_buffer`?

TimosAmpel replied:
This was updated to a queue in the latest implementation. Basically I wanted a way to store all the received data from the VM in queue and exit the process function as soon as possible. At a later time,  the backend thread which writes and reads the tcp_socket consumes them asynchronously.

MatiasVara commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+            return Ok(true);
+        }
+
+        let desc_chain = &requests[0];
```
AFAIU, request is an array of `ConsoleDescriptorChain`. This line means that for reception we are only using the chain at 0, right? why?

TimosAmpel replied:
Hello @MatiasVara, in oldest implementations I couldn't make it work otherwise and this code was left like this since then.
The latest version goes as usual through all the elements into the request array.

TimosAmpel commented on staging/vhost-device-console/README.md:
````diff
@@ -0,0 +1,129 @@
+```text
+host# qemu-system                                                                   \
+    -m 4096                                                                         \
+    -numa node,memdev=mem                                                           \
````
thanks! I just followed vsock example, hope that works!

TimosAmpel commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,129 @@
+
+-- option:: -b, --backend=network|nested
+
+  The backend type vhost-device-console will use.
```
I just saw that both "nested" and "network" are described into the first section in the README file.
Also as you suggested, I have put "nested" as the default backend.

TimosAmpel commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,129 @@
+    -m 4096                                                \
+    -numa node,memdev=mem                                  \
+    -object memory-backend-memfd,id=mem,size=4G,share=on   \
+    -chardev socket,path=/tmp/console.sock,id=con          \
```
Thanks! Indeed that was a mistake, I have just updated it

TimosAmpel commented on staging/vhost-device-console/README.md:
````diff
@@ -0,0 +1,129 @@
+
+The daemon should be started first:
+```shell
+host# vhost-device-console --socket-path=console.sock --socket-count=1 \
````
Thanks! That's corrected

TimosAmpel commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+            .add_used(desc_chain.head_index(), response.as_bytes().len() as u32)
+            .is_err()
+        {
+            warn!("Couldn't return used descriptors to the ring");
```
In most devices as I see they treat it in this way. I liked your idea so in the latest push,
I did as you suggested and `add_used` returns an error if fails.

TimosAmpel commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+                .add_used(desc_chain.head_index(), desc_request.len())
+                .is_err()
+            {
+                warn!("Couldn't return used descriptors to the ring");
```
That was addressed also, by returning an error

TimosAmpel commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+        trace!("value 0x{:x}", control_msg.value.to_native());
+    }
+
+    fn process_rx_requests(
```
The last updated has included the reader/writer logic

TimosAmpel commented on staging/vhost-device-console/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+    socket_path: PathBuf,
+
+    /// Location of vhost-user Unix domain socket. This is suffixed by 0,1,2..socket_count-1.
+    #[clap(short = 'p', long, value_name = "PORT", default_value = "12345")]
```
I have updated the documentation line where you pointed me and described it there a bit more clearly. The README file also describes how that initial value is increased by 1 for each new guest.

TimosAmpel commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+    }
+
+    fn features(&self) -> u64 {
+        // VIRTIO_CONSOLE_F_MULTIPORT is supported but is disabled by default
```
That line is deleted. The `VIRTIO_CONSOLE_F_MULTIPORT` is used by default in the latest version, thanks for the suggestion!

TimosAmpel commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+        vhu_console: &Arc<RwLock<VhostUserConsoleBackend>>,
+    ) -> JoinHandle<Result<()>> {
+        let vhu_console = Arc::clone(vhu_console);
+        trace!("Enter text and press Enter: ");
```
That was left there since the debugging phase and deleted in the latest push

TimosAmpel commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1412 @@
+                                            .into_bytes();
+                                        if let Err(e) = stream.write_all(&byte_stream) {
+                                            eprintln!("Error writing to stream: {}", e);
+                                            panic!("Writing to stream fails");
```
Hmm.. I will have another look to check which of them can be avoided in one of the next weeks

stefano-garzarella replied:
Do you prefer to merge this PR as it is, or check that `panic!()`s ?

TimosAmpel commented:
> After few tries I had this:
> 
> ```
> thread 'vring_worker' panicked at vhost-device-console/src/vhu_console.rs:235:55:
> called `Result::unwrap()` on an `Err` value: FromUtf8Error { bytes: [239], error: Utf8Error { valid_up_to: 0, error_len: None } }
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> thread '<unnamed>' panicked at vhost-device-console/src/vhu_console.rs:560:60:
> called `Result::unwrap()` on an `Err` value: PoisonError { .. }
> ```
> 
> Also I'm not sure if it's releated to the device or the driver, but with telnet or nc, it's not working very well:
> 
> ```
> [root@f38-vm-build ~]# echo hello
> [root@f38-vm-build ~]# echo hello
> 
> [root@f38-vm-build ~]# echo hello
> 
> 
> [root@f38-vm-build ~]# echo hello
> echo hello
> [root@f38-vm-build ~]# 
> ```
> 
> Is that expected? Or am I doing something wrong?

Hmm no, this is not expected. The newest version udpated a lot of things, so the behaviour that you noticed is resolved.
About the error, I did not see anything like this during testing the latest implementation, so I hope this is also fixed.
Please let me know if that's fixed also on your side.

Last but not least, thanks a lot for all the comments / suggestions, I think I have addressed most of them.
Please let me know if you have additional ones.

MatiasVara commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,133 @@
+with the vhost-user frontend devices, and the backend type "nested" or
+"network".
+
+By using the "nested" backend, vhost-device-console, as soon as, the guest
```
This is a minor comment. How about rewrite it as:
```suggestion
The nested backend allows input/output to the guest console through the current terminal.
```

TimosAmpel replied:
Fully agree, nice touch! I just needed to adjuct it a bit for the other backend.

MatiasVara commented on staging/vhost-device-console/README.md:
````diff
@@ -0,0 +1,133 @@
+
+Eventually, the user can connect to the console by running:
+```test
+host# stty raw && nc 127.0.0.1 12345
````
I tested using this cmd line and I observed the following behavior:
```
root@syzkaller:~# llss^M
Connecting  Streaming    file.wav   nada.wav   rec.wav  test.wav
Paused      Unconnected  file2.wav  queen.wav  test.sh
root@syzkaller:~# 
```
Keys are displayed two times. Maybe my configuration is wrong. The command is correctly executed so guest is not getting the key two times.

TimosAmpel replied:
Ouch.. my bad. I have finally got it working with `stty -icanon -echo && nc localhost 12345 && stty echo` and also updated the README accordingly. Please let me know if that solves the issue on your side too.

stefano-garzarella commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1285 @@
+        })
+    }
+    pub fn kill_console_thread(&self) {
+        dbg!("Kill thread\n");
```
I have this output:
```
[vhost-device-console/src/vhu_console.rs:626:9] "Kill thread\n" = "Kill thread\n"
```

```suggestion
        dbg!("Kill thread");
```

Can you also check other `dbg!()`, `trace!()`, etc. calls ?

TimosAmpel replied:
Of course, the latest implementation has deleted all the unnecessarry of the `trace` and `dbg` calls.

stefano-garzarella replied:
Great, but for me it was also fine to leave them if they could be useful. As you prefer ;-)

stefano-garzarella reviewed (commented):
> Hmm no, this is not expected. The newest version udpated a lot of things, so the behaviour that you noticed is resolved.
> About the error, I did not see anything like this during testing the latest implementation, so I hope this is also fixed.
> Please let me know if that's fixed also on your side.

Yep, fixed now. Following the examples in the readme, everything is working as expected.

> 
> Last but not least, thanks a lot for all the comments / suggestions, I think I have addressed most of them.

The one about `panic!()` is missing, but I added a comment there, so we can decide to fix later if you prefer.

> Please let me know if you have additional ones.

CI is complaining about coverage, so I think we can just update the value in the `staging/coverage_config_x86_64.json`.

I left some minor comments, but the PR LGTM. Thanks for the work ;-)

TimosAmpel commented:
> > Hmm no, this is not expected. The newest version udpated a lot of things, so the behaviour that you noticed is resolved.
> > About the error, I did not see anything like this during testing the latest implementation, so I hope this is also fixed.
> > Please let me know if that's fixed also on your side.
> 
> Yep, fixed now. Following the examples in the readme, everything is working as expected.

That's great! 
 
> > Last but not least, thanks a lot for all the comments / suggestions, I think I have addressed most of them.
> 
> The one about `panic!()` is missing, but I added a comment there, so we can decide to fix later if you prefer.

Indeed I missed that about the `panic!()`, in the latest implementation though I have eliminate all of them. In some of those cases, I am just printing an informative error message but the device can continue running, and in the rest I am terminating the current thread (ex. in case of the socket is terminated from the other side).

> > Please let me know if you have additional ones.
> 
> CI is complaining about coverage, so I think we can just update the value in the `staging/coverage_config_x86_64.json`.
Ok done, I have decreased a bit the limit.

> I left some minor comments, but the PR LGTM. Thanks for the work ;-)
Your welcome, thanks you too @stefano-garzarella, @MatiasVara, @epilys, @dorindabassey for your comments / suggestions and overall help!

TimosAmpel commented:
@stefano-garzarella Thanks for the suggections:
```
## Staging Device
This device will be in `staging` until we complete the following steps:
- [ ] merge Linux's driver upstream
- [ ] merge virtio-spec changes upstream
- [ ] ...
```
I did add a some simple critiria, which after implementing them we can consider moving the code into the main directory.
I think at that point I have addressed all your comments / suggestions, please let me know if I missed anything or you have any additional one.

stefano-garzarella commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,137 @@
+
+## Staging Device
+This device will be in `staging` until we complete the following steps:
+- [ ] Enable VIRTIO_CONSOLE_F_SIZE feature
+- [ ] Enable VIRTIO_CONSOLE_F_EMERG_WRITE feature
```
What about  `Support` instead of `Enable` ?

TimosAmpel replied:
Agree, done!

stefano-garzarella commented on staging/vhost-device-console/src/vhu_console.rs:
```diff
@@ -0,0 +1,1279 @@
+            buffer.extend_from_slice(&ctrl_msg.to_le_bytes());
+
+            if ctrl_msg.event.to_native() == VIRTIO_CONSOLE_PORT_NAME {
+                buffer.extend_from_slice(b"org.test.foo!");
```
What is this fixed string?

I suggest adding a `const` for that with a comment.

TimosAmpel replied:
Alright, I have added it as a `const` with a comment saying that this needs to be updated when `MULTIPORT` feature will be supported for more than one devices.

stefano-garzarella commented on staging/vhost-device-console/README.md:
```diff
@@ -0,0 +1,137 @@
+
+## Staging Device
+This device will be in `staging` until we complete the following steps:
+- [ ] Support VIRTIO_CONSOLE_F_SIZE feature
+- [ ] Support VIRTIO_CONSOLE_F_EMERG_WRITE feature
```
I'm fine to keep the device in staging, but if these features are not mandatory in the spec, we can promote it to the main workspace right now.

I don't have a strong opinion, so it is up to you or someone interested on the device to decide when to move it.

TimosAmpel replied:
Based on the [specs](https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.html#x1-2960005), I understand that the features: `VIRTIO_CONSOLE_F_SIZE`,
`VIRTIO_CONSOLE_F_MULTIPORT` and `VIRTIO_CONSOLE_F_EMERG_WRITE` are NOT mandatory.
So I also think that we could move this device directly on the main workspace.

stefano-garzarella replied:
Maybe one of the reason to keep it in staging is the coverage. Ideally, we would like to avoid decreasing a lot the coverage of the main workspace. So if adding tests it will require some time, maybe we can merge this in staging for now, so you can work on tests in the meantime.

WDYT?

TimosAmpel replied:
Sounds good to me, fully agree on that!

stefano-garzarella replied:
@TimosAmpel cool, so can you add something like "Increase test coverage" in this list?
(I hope it's a last change I ask ;-P)

TimosAmpel replied:
No worries at all, I have that added in the list and mark the rest of the bullets as "optional".

MatiasVara commented:
I tested and it seems working fine.

stefano-garzarella commented:
> I tested and it seems working fine.

@MatiasVara thanks for reviewing it!

@stsquad @epilys @vireshk I think we can merge this PR.
